README.md
pushwechat![travis-ci](https://secure.travis-ci.org/dead-horse/weixin-push.png)
====== 

 通过模拟后台登录的形式，进行消息的发送和获取。   

## 用法  

``` js
var Pusher = require('pwechat');
//token可以在登录微信公众平台后台的URL中看到（URL中可以看到token=xxxx）。
var pusher = Pusher.create('youremail', 'yourpassword', 'yourtoken');

pusher.on('PWwchatError', function (err) {
  console.log(err); //无法登录后台
});

/**
 * 给单个用户发送微信消息
 * @param {String} fakeId 用户fakeId
 * @param {String} content 发送内容
 */
pusher.singleSend('12345, '测试内容', function (err, data) {
  // 发送成功的响应data.should.eql({ret: 0, msg: 'ok'});
});

/**
 * 获取包含关键字的消息
 * @param {String} keyword 消息中包含的关键字
 * @param {Number} count 获取条数
 * @param {Number} fromMsgId 从这个msgId开始往前查找
 */
pusher.getMessage('@help', 10, 1000, function (err, data) {
  // 获取成功的响应，data会是一个数组
});
```


## 安装  

```
npm install pwechat
```  

## Lincense  
(The MIT License)

Copyright (c) 2012 dead-horse and other contributors

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



History.md

0.0.4 / 2013-05-03 
==================

  * fix login token 

0.0.1  
==================  

  * Initial release 
  * add getMessage
  * add singleSend 


test/proxy.test.js
/*!
 * pushwechat - lib/proxy.js 
 * Copyright(c) 2012 Taobao.com
 * Author: dead-horse <dead_horse@qq.com>
 */


/**
 * Module dependencies.
 */
var proxy = require('../').proxy;
var mm = require('mm');
var should = require('should');
var urllib = require('urllib');

describe('lib/proxy.js', function () {
  describe('#login', function () {
    afterEach(mm.restore);
    it('should response response error', function (done) {
      mm.data(urllib, 'request', '{a: 1}');
      proxy.login('test@gmail.com', '123', function (err, data) {
        err.message.should.equal('response error');
        should.not.exist(data);
        done();
      });
    });

    it('should response login error', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(new Error('mock error'), '{"a": "b"}');
        });
      });
      proxy.login('test@gmail.com', '123', function (err, data) {
        err.message.should.equal('mock error');
        should.not.exist(data);
        done();
      });      
    });

    it('should response ok', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, '{"ErrCode": 0}', {statusCode: 200, headers: {'set-cookie': ['a=b;'] }});
        }); 
      });
      proxy.login('test@gmail.com', '123', function (err, data) {
        should.not.exist(err);
        data.should.equal('a=b;');
        done();
      });
    });
  });

  describe('#singleSend', function () {
    afterEach(mm.restore);

    it('should reponse err', function (done) {
      mm.error(urllib, 'request', 'mock error');
      proxy.singleSend('123', 'test', 'cookie', 'token', function (err) {
        err.message.should.equal('mock error');
        done();
      });
    });

    it('should response error by statusCode', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, 'test', {statusCode: 302});
        });
      });
      proxy.singleSend('123', 'test', 'cookie', 'token', function (err) {
        err.message.should.equal('singleSend Error! status code:302');
        done();
      });
    });

    it('should response parse error', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, 'test', {statusCode: 200});
        });
      });
      proxy.singleSend('123', 'test', 'cookie', 'token', function (err) {
        err.message.should.equal('singleSend Error! Unexpected token e');
        done();
      });      
    });

    it('should ok', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, '{"ret":0, "msg":"ok"}', {statusCode: 200});
        });
      });
      proxy.singleSend('123', 'test', 'cookie', 'token', function (err, data) {
        should.not.exist(err);
        data.should.eql({ret: 0, msg: 'ok'});
        done();
      });      
    });
  });

  describe('#refresh', function () {
    afterEach(mm.restore);
    
    it('should reponse err', function (done) {
      mm.error(urllib, 'request', 'mock error');
      proxy.refresh('cookie', function (err) {
        err.message.should.equal('mock error');
        done();
      });
    });

    it('should response error by statusCode', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, 'test', {statusCode: 500});
        });
      });
      proxy.refresh('cookie', function (err) {
        err.message.should.equal('refresh Error! status code:500');
        done();
      });
    });

    it('should refresh error by header error', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, '', {statusCode: 302});
        });
      });
      proxy.refresh('cookie', function (err) {
        err.message.should.equal('Refresh error, can not get token');
        done();
      });
    });

    it('should refresh ok', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, '', {statusCode: 200, headers: {location: 'token=1234324'}});
        });
      });
      proxy.refresh('cookie', function (err, data) {
        should.not.exist(err);
        data.should.equal('1234324');
        done();
      });
    });
  });

  describe('#getMessage', function () {
    it('should reponse err', function (done) {
      mm.error(urllib, 'request', 'mock error');
      proxy.getMessage('cookie', 'token', '123', 'test', 100, function (err) {
        err.message.should.equal('mock error');
        done();
      });
    });

    it('should response error by statusCode', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, 'test', {statusCode: 302});
        });
      });
      proxy.getMessage('cookie', 'token', '123', 'test', 100, function (err) {
        err.message.should.equal('getMessage Error! status code:302');
        done();
      });
    });

    it('should response parse error', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, 'test', {statusCode: 200});
        });
      });
      proxy.getMessage('cookie', 'token', '123', 'test', 100, function (err) {
        err.message.should.equal('getMessage Error! Unexpected token e');
        done();
      });      
    });

    it('should get ok', function (done) {
      mm(urllib, 'request', function (url, opts, callback) {
        process.nextTick(function () {
          callback(null, new Buffer('[]'), {statusCode: 200});
        });
      });
      proxy.getMessage('cookie', 'token', '123', 'test', 100, function (err, data) {
        should.not.exist(err);
        data.should.eql([]);
        done();
      });
    });    
  });
});

test/pushwechat.test.js
/*!
 * pushwechat - lib/proxy.js 
 * Copyright(c) 2012 Taobao.com
 * Author: dead-horse <dead_horse@qq.com>
 */


/**
 * Module dependencies.
 */
var proxy = require('../').proxy;
var mm = require('mm');
var should = require('should');
var Pusher = require('../');

var pusher;
var loginBack;
describe('lib/pushwechat.js', function () {
  before(function (done) {
    pusher = Pusher.create('testname', 'testpwd', 'token');
    loginBack = proxy.login;
    proxy.login = function (name, pwd, callback) {
      process.nextTick(function () {
        return callback(null, 'mock cookie');
      });
    }
    pusher.once('connect', done);
  });
  after(function () {
    proxy.login = loginBack;
  });

  describe('#singleSend', function () {
    afterEach(mm.restore);
    
    it('should error when proxy error', function (done) {
      mm.error(proxy, 'singleSend', 'mock error');
      pusher.singleSend('fakeId', 'content', function (err, data) {
        err.message.should.equal('mock error');
        should.not.exist(pusher.cookie);
        pusher.cookie = 'mock cookie';
        done();
      });
    });
    
    it('should ok when has cookie', function (done) {
      mm.data(proxy, 'singleSend', {ret: 0, msg: 'ok'});
      pusher.singleSend('fakeId', 'content', function (err, data) {
        should.not.exist(err);
        data.should.eql({ret: 0, msg: 'ok'});
        done();
      });
    });

    it('should ok when no cookie', function (done) {
      mm.data(proxy, 'singleSend', {ret: 0, msg: 'ok'});
      pusher.cookie = null;
      pusher.singleSend('fakeId', 'content', function (err, data) {
        should.not.exist(err);
        data.should.eql({ret: 0, msg: 'ok'});
        pusher.cookie.should.equal('mock cookie');
        done();
      });
    });
  });

  describe('#getMessage', function () {
    afterEach(mm.restore);

    it('should error when proxy error', function (done) {
      mm.error(proxy, 'getMessage', 'mock error');
      pusher.getMessage('fakeId', 'content', 100, function (err, data) {
        err.message.should.equal('mock error');
        should.not.exist(pusher.cookie);
        pusher.cookie = 'mock cookie';
        done();
      });
    });

    it('should ok when has cookie', function (done) {
      mm.data(proxy, 'getMessage', [{}]);
      pusher.getMessage('fakeId', 'content', 100, function (err, data) {
        should.not.exist(err);
        data.should.eql([{}]);
        done();
      });
    });

    it('should error when no cookie and login error', function (done) {
      pusher.cookie = null;
      mm.error(proxy, 'login', 'mock error');
      pusher.getMessage('fakeId', 'content', 100, function (err, data) {
        err.message.should.equal('mock error');
        should.not.exist(pusher.cookie);
        pusher.cookie = 'mock cookie';
        done();
      });
    });

    it('should ok when no cookie', function (done) {
      mm.data(proxy, 'getMessage', [{}]);
      pusher.cookie = null;
      pusher.getMessage('fakeId', 'content', 100, function (err, data) {
        should.not.exist(err);
        data.should.eql([{}]);
        pusher.cookie.should.equal('mock cookie');
        done();
      });
    });
  });

  describe('#_refresh', function () {
    afterEach(mm.restore);
    it('should refresh ok', function (done) {
      pusher._refresh();
      pusher.once('connect', done);
    });

    it('should refresh error', function (done) {
      mm.error(proxy, 'refresh', 'mock error');
      mm.error(proxy, 'login', 'mock error');
      pusher.once('PWechatError', function (err) {
        err.message.should.equal('mock error');
        should.not.exist(pusher.cookie);
        done();
      });
      pusher._refresh();
    })
  });
});

lib/pushwechat.js
/*!
 * pushwechat - lib/pushwechat.js 
 * Copyright(c) 2012 Taobao.com
 * Author: dead-horse <dead_horse@qq.com>
 */


/**
 * Module dependencies.
 */
var proxy = require('./proxy');
var EventEmitter = require('events').EventEmitter;
var util = require('util');
var utility = require('utility');

var Pusher = function (username, pwd) {
  this.username = username || '';
  this.pwd = pwd || '';
  if (!this.username || !this.pwd) {
    throw new Error('Must have username, password and token');
  }
  EventEmitter.call(this);
  setInterval(this._refresh.bind(this), 10 * 60 * 1000);
  this._login(this._refresh.bind(this));
};
util.inherits(Pusher, EventEmitter);

Pusher.prototype._login = function(callback) {
  var self = this;
  proxy.login(self.username, self.pwd, function (err, cookie, token) {
    if (err) {
      return callback(err);
    }
    self.cookie = cookie;
    self.token = token;
    callback && callback(null, cookie, token);
  });
};

Pusher.prototype.afterLogin = function (callback, args) {
  var self = this;
  this._login(function (err, cookie) {
    if (err) {
      var cb = args[args.length - 1];
      return typeof cb === 'function' && cb(err);
    }
    callback.apply(self, args);
  });
};

Pusher.prototype._refresh = function () {
  var self = this;
  proxy.refresh(this.cookie, function (err, token) {
    if (err) {
      return self._login(function (err) {
        if (err) {
          self.cookie = null;
          self.token = null;
          self.emit('PWechatError', err);
        } else {
          self.emit('connect');          
        }
      });
    }
    self.token = token;
  });
};

/**
 * 给单个用户发送微信消息
 * @param {String} fakeId 用户fakeId
 * @param {String} content 发送内容
 */
Pusher.prototype.singleSend = function (fakeId, content, callback) {
  var self = this;
  if (self.cookie) {
    return proxy.singleSend(fakeId, content, self.cookie, self.token, function (err, data) {
      if (err) {
        self.cookie = null;
      }
      callback(err, data);
    });
  }
  self.afterLogin(self.singleSend, arguments);
};

/**
 * 获取包含关键字的消息
 * @param {String} keyword 消息中包含的关键字
 * @param {Number} count 获取条数
 * @param {Number} fromMsgId 从这个msgId开始往前查找
 */
Pusher.prototype.getMessage = function (keyword, count, fromMsgId, callback) {
  var self = this;
  if (self.cookie) {
    return proxy.getMessage(self.cookie, self.token, keyword, count, fromMsgId, function (err, data) {
      if (err) {
        self.cookie = null;
      }
      callback(err, data);
    });      
  }
  self.afterLogin(self.getMessage, arguments);
};

exports.create = function (username, pwd, token) {
  return new Pusher(username, pwd, token);
};



lib/proxy.js
/*!
 * pushwechat - lib/proxy.js 
 * Copyright(c) 2012 Taobao.com
 * Author: dead-horse <dead_horse@qq.com>
 */


/**
 * Module dependencies.
 */
var urllib = require('urllib');
var utility = require('utility');
var translate = require('./translate');

var URLS = {
  login: 'http://mp.weixin.qq.com/cgi-bin/login?lang=zh_CN',
  addGroup: 'http://mp.weixin.qq.com/cgi-bin/modifygroup?t=ajax-friend-group',
  singleSend: 'http://mp.weixin.qq.com/cgi-bin/singlesend?t=ajax-response&lang=zh_CN',
  refresh: 'http://mp.weixin.qq.com/cgi-bin/indexpage',
  getMessage: 'http://mp.weixin.qq.com/cgi-bin/getmessage?t=ajax-message'
};

var USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_2) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.95 Safari/537.11';

var okStatus = [65201, 65202, 0];
exports.login = function (username, pwd, callback) {
  var data = {
    username: username,
    pwd: utility.md5(pwd.substr(0, 16)).toLowerCase(),
    imgcode: '',
    f: 'json'
  };
  urllib.request(URLS.login, {
    type: 'POST',
    data: data
  }, function (err, data, res) {
    try {
      data = JSON.parse(data);    
    } catch (err) {
      return callback(new Error('response error'));
    }
    if (err || (res.statusCode !== 200 && res.statusCode !== 302) || okStatus.indexOf(data.ErrCode) < 0) {
      return callback(err || new Error('login error!' + translate.loginErrorMap(data.ErrCode)));
    }
    var cookies = '';
    var token;
    if (data.ErrMsg) {
      var matchs = data.ErrMsg.match(/token=(\d+)/);
      token = matchs ? matchs[1] : '';
    }
    var cookieReg = /^([\w_]+)=([\w_=]+);/;
    res.headers['set-cookie'].forEach(function (cookie) {
      var cmatchs = cookie.match(cookieReg);
      if (cmatchs) {
        cookies += cmatchs[0];
      }
    });
    callback(null, cookies, token);
  });
};

exports.singleSend = function (fakeId, content, cookie, token, callback) {
  var data = {
    type: 1,
    content: content,
    tofakeid: fakeId,
    ajax: 1,
    token: token,
    error: false
  };
  var headers = {
    cookie: cookie,
    'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
    Referer: 'http://mp.weixin.qq.com/cgi-bin/singlemsgpage',
  };
  urllib.request(URLS.singleSend, {
    type: 'POST',
    headers: headers,
    data: data
  }, function (err, data, res) {
    if (err || res.statusCode !== 200) {
      return callback(err || new Error('singleSend Error! status code:' + res.statusCode));
    }
    try {
      data = JSON.parse(data);
    } catch (err) {
      return callback(new Error('singleSend Error! ' + err.message));
    }
    callback(null, data);
  });
};

exports.refresh = function (cookie, callback) {
  urllib.request(URLS.refresh, {headers: {cookie: cookie}}, function (err, data, res) {
    if (err || (res.statusCode !== 302 && res.statusCode !== 200)) {
      return callback(err || new Error('refresh Error! status code:' + res.statusCode));
    }
    var token;
    if (res.headers && res.headers.location) {
      var matchs = res.headers.location.match(/token=(\d+)/);
      token = matchs ? matchs[1] : '';
    }
    return callback(token ? null : new Error('Refresh error, can not get token'), token);
  });
};

exports.getMessage = function (cookie, token, keyword, count, fromMsgId, callback) {
  var data = {
    keyword: keyword,
    count: count,
    frommsgid: fromMsgId,
    token: token,
    ajax: 1
  };
  var headers = {
    'User-Agent': USER_AGENT,
    Referer: 'https://mp.weixin.qq.com/cgi-bin/getmessage?token=1811982400&t=wxm-message&lang=zh_CN&count=10&keyword=%E6%B5%8B%E8%AF%95',
    Cookie: cookie
  };
  urllib.request(URLS.getMessage, {
    type: 'POST',
    headers: headers,
    data: data
  }, function (err, data, res) {
    if (err || res.statusCode !== 200) {
      return callback(err || new Error('getMessage Error! status code:' + res.statusCode));
    }
    try {
      data = JSON.parse(data);
    } catch (err) {
      return callback(new Error('getMessage Error! ' + err.message));
    }
    callback(null, data);
  });
};
